File: sdk/src/main/java/com/libreshockwave/chunks/ScriptChunk.java
package com.libreshockwave.chunks;

import com.libreshockwave.format.ChunkType;
import com.libreshockwave.io.BinaryReader;
import com.libreshockwave.lingo.Opcode;

import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

/**
 * Script bytecode chunk (Lscr).
 * Contains the compiled Lingo bytecode for a script.
 */
public record ScriptChunk(
    int id,
    ScriptType scriptType,
    int behaviorFlags,
    List<Handler> handlers,
    List<LiteralEntry> literals,
    List<PropertyEntry> properties,
    List<GlobalEntry> globals,
    byte[] rawBytecode
) implements Chunk {

    @Override
    public ChunkType type() {
        return ChunkType.Lscr;
    }

    public enum ScriptType {
        MOVIE_SCRIPT(1),
        BEHAVIOR(3),
        PARENT(7),
        UNKNOWN(-1);

        private final int code;

        ScriptType(int code) {
            this.code = code;
        }

        public static ScriptType fromCode(int code) {
            for (ScriptType type : values()) {
                if (type.code == code) return type;
            }
            return UNKNOWN;
        }
    }

    public record Handler(
        int nameId,
        int handlerVectorPos,
        int bytecodeLength,
        int bytecodeOffset,
        int argCount,
        int localCount,
        int globalsCount,
        int lineCount,
        List<Integer> argNameIds,
        List<Integer> localNameIds,
        List<Instruction> instructions
    ) {
        public record Instruction(
            int offset,
            Opcode opcode,
            int rawOpcode,
            int argument
        ) {
            @Override
            public String toString() {
                if (rawOpcode >= 0x40) {
                    return String.format("[%d] %s %d", offset, opcode.getMnemonic(), argument);
                } else {
                    return String.format("[%d] %s", offset, opcode.getMnemonic());
                }
            }
        }
    }

    public record LiteralEntry(
        int type,
        int offset,
        Object value
    ) {}

    public record PropertyEntry(
        int nameId
    ) {}

    public record GlobalEntry(
        int nameId
    ) {}

    public Handler findHandler(String name, ScriptNamesChunk names) {
        int nameId = names.findName(name);
        if (nameId < 0) return null;

        for (Handler h : handlers) {
            if (h.nameId == nameId) {
                return h;
            }
        }
        return null;
    }

    public static ScriptChunk read(BinaryReader reader, int id, int version, boolean capitalX) {
        // Lingo scripts are ALWAYS big endian regardless of file byte order
        reader.setOrder(ByteOrder.BIG_ENDIAN);

        // Header - skip first 8 bytes
        reader.seek(8);

        int totalLength = reader.readI32();
        int totalLength2 = reader.readI32();
        int headerLength = reader.readU16();
        int scriptNumber = reader.readU16();

        // Seek to scriptBehavior at offset 38
        reader.seek(38);
        int behaviorFlags = reader.readI32();

        // Seek to handler data at offset 50
        reader.seek(50);
        int handlerVectorsCount = reader.readU16();
        int handlerVectorsOffset = reader.readI32();
        int handlerVectorsSize = reader.readI32();
        int propertyCount = reader.readU16();
        int propertiesOffset = reader.readI32();
        int globalCount = reader.readU16();
        int globalsOffset = reader.readI32();
        int handlerInfoCount = reader.readU16();
        int handlersOffset = reader.readI32();
        int literalCount = reader.readU16();
        int literalsOffset = reader.readI32();
        int literalDataLen = reader.readI32();
        int literalDataOffset = reader.readI32();

        int scriptType = scriptNumber; // Actually script number, not type

        // Read properties
        List<PropertyEntry> properties = new ArrayList<>();
        if (propertyCount > 0 && propertiesOffset > 0) {
            reader.setPosition(propertiesOffset);
            for (int i = 0; i < propertyCount; i++) {
                properties.add(new PropertyEntry(reader.readI16()));
            }
        }

        // Read globals
        List<GlobalEntry> globals = new ArrayList<>();
        if (globalCount > 0 && globalsOffset > 0) {
            reader.setPosition(globalsOffset);
            for (int i = 0; i < globalCount; i++) {
                globals.add(new GlobalEntry(reader.readI16()));
            }
        }

        // Read literals
        List<LiteralEntry> literals = new ArrayList<>();
        if (literalCount > 0 && literalsOffset > 0) {
            reader.setPosition(literalsOffset);
            List<int[]> literalInfo = new ArrayList<>();
            for (int i = 0; i < literalCount; i++) {
                int type = reader.readI32();
                int offset = reader.readI32();
                literalInfo.add(new int[]{type, offset});
            }

            // Read literal data
            for (int[] info : literalInfo) {
                int type = info[0];
                int offset = info[1];
                Object value = null;

                reader.setPosition(literalDataOffset + offset);
                int dataLen = reader.readI32();

                switch (type) {
                    case 1 -> { // String (dataLen includes null terminator)
                        String s = reader.readStringMacRoman(dataLen);
                        // Strip null terminator if present
                        if (s.endsWith("\0")) {
                            s = s.substring(0, s.length() - 1);
                        }
                        value = s;
                    }
                    case 4 -> { // Int
                        value = dataLen; // Actually the value
                    }
                    case 9 -> { // Float
                        value = (double) Float.intBitsToFloat(reader.readI32());
                    }
                    default -> {
                        value = reader.readBytes(dataLen);
                    }
                }

                literals.add(new LiteralEntry(type, offset, value));
            }
        }

        // Read handlers
        List<Handler> handlers = new ArrayList<>();
        if (handlerInfoCount > 0 && handlersOffset > 0) {
            reader.setPosition(handlersOffset);

            for (int i = 0; i < handlerInfoCount; i++) {
                // Handler record structure from ProjectorRays handler.cpp:readRecord()
                int nameId = reader.readI16();           // int16
                int handlerVectorPos = reader.readU16(); // uint16
                int bytecodeLen = reader.readI32();      // uint32
                int bytecodeOffset = reader.readI32();   // uint32
                int argCount = reader.readU16();         // uint16
                int argOffset = reader.readI32();        // uint32
                int localCount = reader.readU16();       // uint16
                int localOffset = reader.readI32();      // uint32
                int handlerGlobalsCount = reader.readU16();   // uint16
                int handlerGlobalsOffset = reader.readI32();  // uint32
                int unknown1 = reader.readI32();         // uint32
                int unknown2 = reader.readU16();         // uint16
                int lineCount = reader.readU16();        // uint16
                int lineOffset = reader.readI32();       // uint32
                if (capitalX) {
                    int stackHeight = reader.readI32();  // uint32 (only in LctX)
                }

                int savedPos = reader.getPosition();

                // Read argument names (using unsigned 16-bit per ProjectorRays)
                List<Integer> argNameIds = new ArrayList<>();
                if (argCount > 0 && argOffset > 0) {
                    reader.setPosition(argOffset);
                    for (int j = 0; j < argCount; j++) {
                        argNameIds.add(reader.readU16());
                    }
                }

                // Read local variable names (using unsigned 16-bit per ProjectorRays)
                List<Integer> localNameIds = new ArrayList<>();
                if (localCount > 0 && localOffset > 0) {
                    reader.setPosition(localOffset);
                    for (int j = 0; j < localCount; j++) {
                        localNameIds.add(reader.readU16());
                    }
                }

                // Parse bytecode instructions (matching dirplayer-rs handler.rs)
                List<Handler.Instruction> instructions = new ArrayList<>();
                if (bytecodeLen > 0 && bytecodeOffset > 0) {
                    reader.setPosition(bytecodeOffset);
                    int bytecodeEnd = bytecodeOffset + bytecodeLen;

                    while (reader.getPosition() < bytecodeEnd) {
                        int instrOffset = reader.getPosition() - bytecodeOffset;
                        int op = reader.readU8();

                        // Normalize opcode: multi-byte opcodes are 0x40 + (op % 0x40)
                        Opcode opcode = Opcode.fromCode(op >= 0x40 ? (0x40 + op % 0x40) : op);
                        int argument = 0;

                        // Argument size is determined by the op byte value, not opcode type
                        if (op >= 0xC0) {
                            // 4-byte argument
                            argument = reader.readI32();
                        } else if (op >= 0x80) {
                            // 2-byte argument
                            if (opcode == Opcode.PUSH_INT16 || opcode == Opcode.PUSH_INT8) {
                                // Treat pushint's arg as signed
                                argument = reader.readI16();
                            } else {
                                argument = reader.readU16();
                            }
                        } else if (op >= 0x40) {
                            // 1-byte argument
                            if (opcode == Opcode.PUSH_INT8) {
                                // Treat pushint's arg as signed
                                argument = reader.readI8();
                            } else {
                                argument = reader.readU8();
                            }
                        }
                        // op < 0x40: no argument (single-byte opcode)

                        instructions.add(new Handler.Instruction(instrOffset, opcode, op, argument));
                    }
                }

                reader.setPosition(savedPos);

                handlers.add(new Handler(
                    nameId,
                    handlerVectorPos,
                    bytecodeLen,
                    bytecodeOffset,
                    argCount,
                    localCount,
                    handlerGlobalsCount,
                    lineCount,
                    argNameIds,
                    localNameIds,
                    instructions
                ));
            }
        }

        // Read raw bytecode for reference
        byte[] rawBytecode = new byte[0];

        return new ScriptChunk(
            id,
            ScriptType.fromCode(scriptType),
            behaviorFlags,
            handlers,
            literals,
            properties,
            globals,
            rawBytecode
        );
    }
}

File: cpp/src/chunks/ScriptChunk.cpp
#include "libreshockwave/chunks/ScriptChunk.hpp"

#include <algorithm>
#include <bit>
#include <cctype>
#include <cstddef>
#include <cstdint>
#include <sstream>
#include <utility>

#include "libreshockwave/io/BinaryReader.hpp"
#include "libreshockwave/chunks/ScriptNamesChunk.hpp"

namespace libreshockwave::chunks {
namespace {

class ScopedByteOrder {
public:
    ScopedByteOrder(io::BinaryReader& reader, io::ByteOrder order)
        : reader_(reader), originalOrder_(reader.order()) {
        reader_.setOrder(order);
    }

    ~ScopedByteOrder() {
        reader_.setOrder(originalOrder_);
    }

private:
    io::BinaryReader& reader_;
    io::ByteOrder originalOrder_;
};

bool canReadAt(const io::BinaryReader& reader, int offset, std::size_t length) {
    if (offset < 0) {
        return false;
    }
    const auto start = static_cast<std::size_t>(offset);
    return start <= reader.length() && length <= reader.length() - start;
}

std::string floatingToString(double value) {
    std::ostringstream out;
    out << value;
    return out.str();
}

std::vector<int> readNameIds(io::BinaryReader& reader, int count, int offset) {
    std::vector<int> ids;
    if (count <= 0 || !canReadAt(reader, offset, static_cast<std::size_t>(count) * 2U)) {
        return ids;
    }

    const auto savedPosition = reader.position();
    reader.seek(static_cast<std::size_t>(offset));
    ids.reserve(static_cast<std::size_t>(count));
    for (int index = 0; index < count; ++index) {
        ids.push_back(reader.readU16());
    }
    reader.seek(savedPosition);
    return ids;
}

bool equalsIgnoreCase(std::string_view lhs, std::string_view rhs) {
    if (lhs.size() != rhs.size()) {
        return false;
    }
    for (std::size_t index = 0; index < lhs.size(); ++index) {
        if (std::tolower(static_cast<unsigned char>(lhs[index])) !=
            std::tolower(static_cast<unsigned char>(rhs[index]))) {
            return false;
        }
    }
    return true;
}

} // namespace

ScriptChunkType scriptChunkTypeFromCode(int code) {
    switch (code) {
        case 1: return ScriptChunkType::Score;
        case 2: return ScriptChunkType::Behavior;
        case 3: return ScriptChunkType::MovieScript;
        case 7: return ScriptChunkType::Parent;
        default: return ScriptChunkType::Unknown;
    }
}

std::string ScriptChunk::Instruction::toString() const {
    std::ostringstream out;
    out << "[" << offset << "] " << lingo::mnemonic(opcode);
    if (rawOpcode >= 0x40) {
        out << " " << argument;
    }
    return out.str();
}

int ScriptChunk::Handler::getInstructionIndex(int offset) const {
    if (const auto found = bytecodeIndexMap.find(offset); found != bytecodeIndexMap.end()) {
        return found->second;
    }
    return -1;
}

ScriptChunk::ScriptChunk(const DirectorFile* file,
                         id::ChunkId id,
                         ScriptChunkType scriptType,
                         int behaviorFlags,
                         std::vector<Handler> handlers,
                         std::vector<LiteralEntry> literals,
                         std::vector<PropertyEntry> properties,
                         std::vector<GlobalEntry> globals,
                         std::vector<std::uint8_t> rawBytecode)
    : file_(file),
      id_(id),
      scriptType_(scriptType),
      behaviorFlags_(behaviorFlags),
      handlers_(std::move(handlers)),
      literals_(std::move(literals)),
      properties_(std::move(properties)),
      globals_(std::move(globals)),
      rawBytecode_(std::move(rawBytecode)) {}

const DirectorFile* ScriptChunk::file() const { return file_; }
format::ChunkType ScriptChunk::type() const { return format::ChunkType::Lscr; }
id::ChunkId ScriptChunk::id() const { return id_; }
ScriptChunkType ScriptChunk::scriptType() const { return scriptType_; }
int ScriptChunk::behaviorFlags() const { return behaviorFlags_; }
const std::vector<ScriptChunk::Handler>& ScriptChunk::handlers() const { return handlers_; }
const std::vector<ScriptChunk::LiteralEntry>& ScriptChunk::literals() const { return literals_; }
const std::vector<ScriptChunk::PropertyEntry>& ScriptChunk::properties() const { return properties_; }
const std::vector<ScriptChunk::GlobalEntry>& ScriptChunk::globals() const { return globals_; }
const std::vector<std::uint8_t>& ScriptChunk::rawBytecode() const { return rawBytecode_; }
bool ScriptChunk::hasProperties() const { return !properties_.empty(); }
bool ScriptChunk::hasGlobals() const { return !globals_.empty(); }

std::optional<ScriptChunk::Handler> ScriptChunk::findHandlerByNameId(int nameId) const {
    for (const auto& handler : handlers_) {
        if (handler.nameId == nameId) {
            return handler;
        }
    }
    return std::nullopt;
}

std::string ScriptChunk::getHandlerName(const Handler& handler, const ScriptNamesChunk* names) const {
    if (names) {
        return names->getName(handler.nameId);
    }
    return "handler#" + std::to_string(handler.nameId);
}

std::string ScriptChunk::resolveName(int nameId, const ScriptNamesChunk* names) const {
    if (names) {
        return names->getName(nameId);
    }
    return "#" + std::to_string(nameId);
}

std::optional<ScriptChunk::Handler> ScriptChunk::findHandler(std::string_view name, const ScriptNamesChunk* names) const {
    if (!names) {
        return std::nullopt;
    }
    for (const auto& handler : handlers_) {
        if (equalsIgnoreCase(names->getName(handler.nameId), name)) {
            return handler;
        }
    }
    return std::nullopt;
}

std::vector<std::string> ScriptChunk::getPropertyNames(const ScriptNamesChunk* names) const {
    std::vector<std::string> result;
    if (!names) {
        return result;
    }
    result.reserve(properties_.size());
    for (const auto& property : properties_) {
        result.push_back(names->getName(property.nameId));
    }
    return result;
}

std::vector<std::string> ScriptChunk::getGlobalNames(const ScriptNamesChunk* names) const {
    std::vector<std::string> result;
    if (!names) {
        return result;
    }
    result.reserve(globals_.size());
    for (const auto& global : globals_) {
        result.push_back(names->getName(global.nameId));
    }
    return result;
}

ScriptChunk ScriptChunk::read(const DirectorFile* file,
                              io::BinaryReader& reader,
                              id::ChunkId chunkId,
                              int version,
                              bool capitalX) {
    ScopedByteOrder order(reader, io::ByteOrder::BigEndian);

    int scriptNumber = 0;
    int behaviorFlags = 0;
    int handlerVectorsCount = 0;
    int handlerVectorsOffset = 0;
    int handlerVectorsSize = 0;
    int propertyCount = 0;
    int propertiesOffset = 0;
    int globalCount = 0;
    int globalsOffset = 0;
    int handlerInfoCount = 0;
    int handlersOffset = 0;
    int literalCount = 0;
    int literalsOffset = 0;
    int literalDataLen = 0;
    int literalDataOffset = 0;

    if (canReadAt(reader, 8, 12)) {
        reader.seek(8);
        (void)reader.readI32();
        (void)reader.readI32();
        (void)reader.readU16();
        scriptNumber = reader.readU16();
    }

    if (canReadAt(reader, 38, 4)) {
        reader.seek(38);
        behaviorFlags = reader.readI32();
    }

    int scriptTypeCode = behaviorFlags & 0x0F;
    if (scriptTypeCode == 0) {
        scriptTypeCode = scriptNumber;
    }

    if (canReadAt(reader, 50, 42)) {
        reader.seek(50);
        handlerVectorsCount = reader.readU16();
        handlerVectorsOffset = reader.readI32();
        handlerVectorsSize = reader.readI32();
        propertyCount = reader.readU16();
        propertiesOffset = reader.readI32();
        globalCount = reader.readU16();
        globalsOffset = reader.readI32();
        handlerInfoCount = reader.readU16();
        handlersOffset = reader.readI32();
        literalCount = reader.readU16();
        literalsOffset = reader.readI32();
        literalDataLen = reader.readI32();
        literalDataOffset = reader.readI32();
    }
    (void)handlerVectorsCount;
    (void)handlerVectorsOffset;
    (void)handlerVectorsSize;
    (void)literalDataLen;

    std::vector<PropertyEntry> properties;
    if (propertyCount > 0 && canReadAt(reader, propertiesOffset, static_cast<std::size_t>(propertyCount) * 2U)) {
        reader.seek(static_cast<std::size_t>(propertiesOffset));
        properties.reserve(static_cast<std::size_t>(propertyCount));
        for (int index = 0; index < propertyCount; ++index) {
            properties.push_back(PropertyEntry{reader.readI16()});
        }
    }

    std::vector<GlobalEntry> globals;
    if (globalCount > 0 && canReadAt(reader, globalsOffset, static_cast<std::size_t>(globalCount) * 2U)) {
        reader.seek(static_cast<std::size_t>(globalsOffset));
        globals.reserve(static_cast<std::size_t>(globalCount));
        for (int index = 0; index < globalCount; ++index) {
            globals.push_back(GlobalEntry{reader.readI16()});
        }
    }

    std::vector<LiteralEntry> literals;
    const int literalRecordLen = version < 0x4B1 ? 6 : 8;
    if (literalCount > 0 && canReadAt(reader, literalsOffset, static_cast<std::size_t>(literalCount * literalRecordLen))) {
        reader.seek(static_cast<std::size_t>(literalsOffset));

        struct LiteralInfo {
            int type;
            int offset;
        };
        std::vector<LiteralInfo> literalInfo;
        literalInfo.reserve(static_cast<std::size_t>(literalCount));
        for (int index = 0; index < literalCount; ++index) {
            const int literalType = version < 0x4B1 ? reader.readU16() : reader.readI32();
            const int offset = reader.readI32();
            literalInfo.push_back(LiteralInfo{literalType, offset});
        }

        literals.reserve(literalInfo.size());
        for (const auto& info : literalInfo) {
            LiteralValue value;
            double numericValue = 0.0;

            if (info.type == 4) {
                value = info.offset;
            } else if (canReadAt(reader, literalDataOffset + info.offset, 4)) {
                reader.seek(static_cast<std::size_t>(literalDataOffset + info.offset));
                const int dataLen = reader.readI32();
                if (dataLen >= 0 && static_cast<std::size_t>(dataLen) <= reader.bytesLeft()) {
                    switch (info.type) {
                        case 1: {
                            std::string text = reader.readStringMacRoman(static_cast<std::size_t>(dataLen));
                            if (!text.empty() && text.back() == '\0') {
                                text.pop_back();
                            }
                            value = std::move(text);
                            break;
                        }
                        case 9: {
                            if (reader.bytesLeft() >= 4) {
                                const auto bits = static_cast<std::uint32_t>(reader.readI32());
                                numericValue = static_cast<double>(std::bit_cast<float>(bits));
                                value = floatingToString(numericValue);
                            }
                            break;
                        }
                        default:
                            value = reader.readBytes(static_cast<std::size_t>(dataLen));
                            break;
                    }
                }
            }

            literals.push_back(LiteralEntry{info.type, info.offset, std::move(value), numericValue});
        }
    }

    std::vector<Handler> handlers;
    const int handlerRecordLen = capitalX ? 46 : 42;
    if (handlerInfoCount > 0 && canReadAt(reader, handlersOffset, static_cast<std::size_t>(handlerInfoCount * handlerRecordLen))) {
        reader.seek(static_cast<std::size_t>(handlersOffset));
        handlers.reserve(static_cast<std::size_t>(handlerInfoCount));

        for (int index = 0; index < handlerInfoCount; ++index) {
            const int nameId = reader.readI16();
            const int handlerVectorPos = reader.readU16();
            const int bytecodeLen = reader.readI32();
            const int bytecodeOffset = reader.readI32();
            const int argCount = reader.readU16();
            const int argOffset = reader.readI32();
            const int localCount = reader.readU16();
            const int localOffset = reader.readI32();
            const int handlerGlobalsCount = reader.readU16();
            (void)reader.readI32();
            (void)reader.readI32();
            (void)reader.readU16();
            const int lineCount = reader.readU16();
            (void)reader.readI32();
            if (capitalX) {
                (void)reader.readI32();
            }

            const auto savedPosition = reader.position();
            auto argNameIds = readNameIds(reader, argCount, argOffset);
            auto localNameIds = readNameIds(reader, localCount, localOffset);

            std::vector<Instruction> instructions;
            std::unordered_map<int, int> bytecodeIndexMap;
            if (bytecodeLen > 0 && canReadAt(reader, bytecodeOffset, static_cast<std::size_t>(bytecodeLen))) {
                reader.seek(static_cast<std::size_t>(bytecodeOffset));
                const auto bytecodeEnd = static_cast<std::size_t>(bytecodeOffset) + static_cast<std::size_t>(bytecodeLen);

                while (reader.position() < bytecodeEnd) {
                    const int instrOffset = static_cast<int>(reader.position()) - bytecodeOffset;
                    const int op = reader.readU8();
                    const auto opcode = lingo::opcodeFromCode(op >= 0x40 ? (0x40 + op % 0x40) : op);
                    int argument = 0;

                    if (op >= 0xC0) {
                        if (reader.position() + 4 > bytecodeEnd) {
                            break;
                        }
                        argument = reader.readI32();
                    } else if (op >= 0x80) {
                        if (reader.position() + 2 > bytecodeEnd) {
                            break;
                        }
                        argument = (opcode == lingo::Opcode::PUSH_INT16 || opcode == lingo::Opcode::PUSH_INT8)
                            ? reader.readI16()
                            : reader.readU16();
                    } else if (op >= 0x40) {
                        if (reader.position() + 1 > bytecodeEnd) {
                            break;
                        }
                        argument = opcode == lingo::Opcode::PUSH_INT8 ? reader.readI8() : reader.readU8();
                    }

                    bytecodeIndexMap[instrOffset] = static_cast<int>(instructions.size());
                    instructions.push_back(Instruction{instrOffset, opcode, op, argument});
                }
            }
            reader.seek(savedPosition);

            handlers.push_back(Handler{nameId,
                                       handlerVectorPos,
                                       bytecodeLen,
                                       bytecodeOffset,
                                       argCount,
                                       localCount,
                                       handlerGlobalsCount,
                                       lineCount,
                                       std::move(argNameIds),
                                       std::move(localNameIds),
                                       std::move(instructions),
                                       std::move(bytecodeIndexMap)});
        }
    }

    return ScriptChunk(file,
                       chunkId,
                       scriptChunkTypeFromCode(scriptTypeCode),
                       behaviorFlags,
                       std::move(handlers),
                       std::move(literals),
                       std::move(properties),
                       std::move(globals),
                       {});
}

} // namespace libreshockwave::chunks
